examples/panel/book-recommender/app.py

```diff
@@ -0,0 +1,132 @@
+"""
+Demonstrates how to use the `ChatInterface` and a `callback` function to respond.
+
+The chatbot Assistant echoes back the message entered by the User.
+
+https://github.com/holoviz-topics/panel-chat-examples/blob/main/docs/examples/basics/basic_chat.py
+"""
+
+import panel as pn
+from openai import OpenAI
+from scipy.spatial import KDTree
+import numpy as np
+
+from rag_book_recommender import get_book_description_by_title, EmbeddingsStore, get_authors, get_embeddings
+
+
+client = OpenAI()
+
+pn.extension()
+
+store = EmbeddingsStore()
+all_authors = get_authors()
+
+
+def detect_author(user_query):
+    system_prompt = f"""
+    You're a system that determines the author in user query. 
+    
+    You need to return only he author name.Please fix any typo if possible
+"""
+    response = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=[
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": "What are some books by Sandra Boynton"},
+            {"role": "system", "content": "Sandra Boynton"},
+            {"role": "user", "content": user_query},
+        ],
+        seed=42,
+        n=1,
+    )
+    author = response.choices[0].message.content.upper()
+    return author if author in all_authors else ""
+
+
+def book_recommender_agent(user_query, verbose=False):
+    """An agent that can retrieve news by topic and summarizes them"""
+    # determine the topic based on the query
+    embeddings_json = get_embeddings()
+    author = detect_author(user_query)
```

Comment:
This approach is pretty restrictive. if I understand correctly, whatever my input is, it'll be mapped to a single author (?) but users might ask something like "show me novels about spies", and there are many authors that would match that criteria.

instead. I suggest this:

- use openai's embedding AI to embed the user query
- then, find the closest matches to the embedding by looking into the book embeddings
- then retrieve the book's info (title, author, description), and display it

Reply:
update: I asked "novels about spies and got this"

![image](https://github.com/ploomber/doc/assets/989250/35d24b74-6147-4994-8680-6dbb71189bea)

which includes two authors, so I'm not sure what the logic is. the code suggests that it should only return books by a single author (?)

Reply:
No it doesn't return books by a single author. Since author is a separate column if user asks recommend books by `some-author` it doesn't work if we only compute similarity with the description field. So if any specific author is mentioned it first filters based on authors and then returns. If no author is mentioned it compares with embeddings of all available books

Reply:
Added a [comment](https://github.com/ploomber/doc/pull/127/files#diff-99d50d468b15d576674758b1fbde220b390f380dc2398dd8b4024a04218dd2ccR89-R90) for clarity